src/main/java/com/google/cloud/pubsublite/flink/internal/sink/MessagePublisher.java

```diff
@@ -22,21 +22,40 @@
 import com.google.cloud.pubsublite.internal.CheckedApiException;
 import com.google.cloud.pubsublite.internal.ExtractStatus;
 import com.google.cloud.pubsublite.internal.Publisher;
+import com.google.cloud.pubsublite.internal.wire.SystemExecutors;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.Semaphore;
 
 public class MessagePublisher implements BulkWaitPublisher<Message> {
   private final Publisher<MessageMetadata> publisher;
   private final List<ApiFuture<MessageMetadata>> publishes;
 
-  public MessagePublisher(Publisher<MessageMetadata> publisher) {
+  private final int maxBytesOutstanding;
+  private final Semaphore bytesOutstanding;
+
+  public MessagePublisher(Publisher<MessageMetadata> publisher, int maxBytesOutstanding) {
     this.publisher = publisher;
     this.publishes = new ArrayList<>();
+    this.maxBytesOutstanding = maxBytesOutstanding;
+    this.bytesOutstanding = new Semaphore(maxBytesOutstanding);
+  }
+
+  private int getAccountedSize(Message message) {
+    long size = message.toProto().getSerializedSize();
+    if (size > maxBytesOutstanding) {
+      return maxBytesOutstanding;
+    }
+    return Math.toIntExact(size);
   }
 
   @Override
-  public void publish(Message message) {
-    publishes.add(publisher.publish(message));
+  public void publish(Message message) throws InterruptedException {
+    final int size = getAccountedSize(message);
+    bytesOutstanding.acquire(size);
```

Comment:
consider adding a tryAcquire first, that logs warning if flow controlled before unconditionally calling acquire?

Reply:
Oh, yeah that's a good idea. Done